tokio-postgres/src/copy_both.rs

```diff
@@ -0,0 +1,280 @@
+use std::marker::{PhantomData, PhantomPinned};
+use std::pin::Pin;
+use std::task::{Context, Poll};
+
+/// This stream is consumed by the Connection once we enter the CopyBoth sub protocol. When the
+/// user drops their DuplexStream, CopyBothReceiver will automatically send a CopyDone or Sync
+/// message to the backend
+pub struct CopyBothReceiver {
+    message_receiver: mpsc::Receiver<FrontendMessage>,
+    error_receiver: oneshot::Receiver<()>,
+    done: bool,
+}
+
+impl CopyBothReceiver {
+    pub(crate) fn new(
+        message_receiver: mpsc::Receiver<FrontendMessage>,
+        error_receiver: oneshot::Receiver<()>,
+    ) -> CopyBothReceiver {
+        CopyBothReceiver {
+            message_receiver,
+            error_receiver,
+            done: false,
+        }
+    }
+}
+
+impl Stream for CopyBothReceiver {
+    type Item = FrontendMessage;
+
+    fn poll_next(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<FrontendMessage>> {
+        if self.done {
+            return Poll::Ready(None);
+        }
+
+        // If an error was received from the backend we have to send a Sync message and not send
+        // any other messages
+        if let Ok(Some(_)) = self.error_receiver.try_recv() {
+            self.message_receiver.close();
+            self.done = true;
+            let mut buf = BytesMut::new();
+            frontend::sync(&mut buf);
+            return Poll::Ready(Some(FrontendMessage::Raw(buf.freeze())));
+        }
+
+        match ready!(self.message_receiver.poll_next_unpin(cx)) {
+            Some(message) => Poll::Ready(Some(message)),
+            None => {
+                self.done = true;
+                let mut buf = BytesMut::new();
+                frontend::copy_done(&mut buf);
+                Poll::Ready(Some(FrontendMessage::Raw(buf.freeze())))
+            }
+        }
+    }
+}
+
+/// The state machine of a CopyBothDuplex
+///
+/// ```ignore
+///    CopyBoth
+///      /  \
+///     v    v
+/// CopyOut CopyIn
+///      \   /
+///       v v
+///   ReadingCopy
+///        |
+///        v
+///     Reading
+/// ```
+#[derive(Clone, Copy)]
+enum DuplexState {
+    /// Initial state where CopyData messages can go in both directions
+    CopyBoth,
+    /// The server->client stream is closed and we're in CopyIn mode
+    CopyIn,
+    /// The client->server stream is closed and we're in CopyOut mode
+    CopyOut,
+    /// Both directions are closed and we're waiting for the first CommandComplete message
+    ReadingCopy,
+    /// Both directions are closed and we're waiting for the final CommandComplete message
+    Reading,
+}
+
+pin_project! {
+    /// A duplex stream for consuming streaming replication data.
+    ///
+    /// The stream side *must* be consumed even if not required in order to process the messages
+    /// coming from the server. If it is not, `Sink::close` may hang forever waiting for the stream
+    /// messages to be consumed.
```

Comment:
Can you expand on the reasoning here? `close()` or `drop()` should certainly ensure that a `CopyDone` is sent. And if the receiver is closed, that should cause future messages received in this sub-protocol to be discarded, eventually allowing the protocol to resume normal operations. Right?